model/routing/ResourceUrlBuilder.php

```diff
@@ -15,81 +15,87 @@
      * @return string
      */
-    public function buildUrl(\core_kernel_classes_Resource $resource)
+    public function buildUrl(core_kernel_classes_Resource $resource)
     {
-        if (!$resource->exists()) {
-            throw new \InvalidArgumentException('The requested resource does not exist or has been deleted');
-        }
+        $resourceClass = $resource->getClass($resource);
 
-        /** @var Perspective $perspective */
-        /** @var Section $section */
+        if (!$resource->exists() && !$resourceClass->exists()) {
+            throw new InvalidArgumentException('The requested resource does not exist or has been deleted');
+        }
 
-        $resourceClass = $resource->isClass()
-            ? $this->getClass($resource)
-            : array_values($resource->getTypes())[0];
+        if (!$resource->isClass()) {
+            $resourceClass = array_values($resource->getTypes())[0];
+        }
 
+        /** @var Perspective $perspective */
         foreach (MenuService::getAllPerspectives() as $perspective) {
+            /** @var Section $section */
             foreach ($perspective->getChildren() as $section) {
                 if ($this->isSectionApplicable($resourceClass, $section)) {
                     return $this->getBackofficeUrl($perspective, $section, $resource);
                 }
             }
         }
 
-        throw new \LogicException('No url could be built for "' . $resource->getUri() . '"');
+        throw new LogicException(
+            sprintf(
+                'No url could be built for "%s"',
+                $resource->getUri()
+            )
+        );
     }
 
     /**
      * Generates the actual URL based on perspective and section
-     * @param Perspective $perspective
-     * @param Section $section
-     * @param \core_kernel_classes_Resource $resource
+     *
+     * @return string
      */
-    private function getBackofficeUrl(Perspective $perspective, Section $section, \core_kernel_classes_Resource $resource)
-    {
+    private function getBackofficeUrl(
+        Perspective $perspective,
+        Section $section,
+        core_kernel_classes_Resource $resource
+    ) {
         return _url('index', 'Main', 'tao', [
             'structure' => $perspective->getId(),
             'section' => $section->getId(),
             'uri' => $resource->getUri(),
         ]);
     }
 
-    /**
-     * @param \core_kernel_classes_Class $resourceClass
-     * @param Section                    $section
-     * @return bool
-     */
-    private function isSectionApplicable(\core_kernel_classes_Class $resourceClass, Section $section)
+    private function isSectionApplicable(core_kernel_classes_Class $resourceClass, Section $section): bool
     {
         /** @var Tree $tree */
         foreach ($section->getTrees() as $tree) {
-            $rootClass = $this->getClass($tree->get('rootNode'));
+            $rootClass = $resourceClass->getClass($tree->get('rootNode'));
```

Comment:
This makes this not reliable according to the previous code https://github.com/oat-sa/extension-tao-backoffice/pull/144/files#diff-4083f9ee23659bc29492c61f9b2ba198c83d6dd7689a79873b8310e96c6ddd6dR54